homeassistant/components/datetime/__init__.py

```diff
@@ -0,0 +1,177 @@
+"""Component to allow setting date/time as platforms."""
+from __future__ import annotations
+
+from dataclasses import dataclass
+from datetime import datetime, timedelta, timezone, tzinfo
+import functools
+import logging
+from typing import final
+
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import ATTR_DATE, ATTR_TIME
+from homeassistant.core import HomeAssistant, ServiceCall
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.config_validation import (  # noqa: F401
+    ENTITY_SERVICE_FIELDS,
+    PLATFORM_SCHEMA,
+    PLATFORM_SCHEMA_BASE,
+)
+from homeassistant.helpers.entity import Entity, EntityDescription
+from homeassistant.helpers.entity_component import EntityComponent
+from homeassistant.helpers.typing import ConfigType
+from homeassistant.util import dt as dt_util
+
+from .const import ATTR_DATETIME, ATTR_TIME_ZONE, DOMAIN, SERVICE_SET_VALUE
+
+SCAN_INTERVAL = timedelta(seconds=30)
+
+ENTITY_ID_FORMAT = DOMAIN + ".{}"
+
+_LOGGER = logging.getLogger(__name__)
+
+__all__ = ["DOMAIN", "DateTimeEntity", "DateTimeEntityDescription"]
+
+ATTR_OFFSET = "offset"
+
+
+def _split_date_time(config):
+    """Split date/time components."""
+    datetime_ = config.pop(ATTR_DATETIME, None)
+    config[ATTR_DATE] = datetime_.date()
+    config[ATTR_TIME] = datetime_.time()
+    if (offset := datetime_.utcoffset()) is not None:
+        config[ATTR_OFFSET] = offset
+    return config
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up Date/Time entities."""
+    component = hass.data[DOMAIN] = EntityComponent[DateTimeEntity](
+        _LOGGER, DOMAIN, hass, SCAN_INTERVAL
+    )
+    await component.async_setup(config)
+
+    component.async_register_entity_service(
+        SERVICE_SET_VALUE,
+        vol.All(
+            vol.Any(
+                vol.All(
+                    vol.Schema(
+                        {
+                            vol.Optional(ATTR_DATE): cv.date,
+                            vol.Optional(ATTR_TIME): cv.time,
+                            vol.Optional(ATTR_TIME_ZONE): cv.time_zone,
+                            **ENTITY_SERVICE_FIELDS,
```

Comment:
The frontend PR has been adjusted by @bramkragten.
These splits isn't needed anymore now, so we can just keep it at accepting a single `datetime` field.

Let's remove the split.